sympy/ntheory/residue_ntheory.py

```diff
@@ -578,45 +546,37 @@ def sqrt_mod_iter(a, p, domain=int):
     >>> from sympy.ntheory.residue_ntheory import sqrt_mod_iter
     >>> list(sqrt_mod_iter(11, 43))
     [21, 22]
+
+    See Also
+    ========
+
+    sqrt_mod : Same functionality, but you want a sorted list or only one solution.
+
     """
     a, p = as_int(a), abs(as_int(p))
-    if isprime(p):
-        a = a % p
-        if a == 0:
-            res = _sqrt_mod1(a, p, 1)
+    v = []
+    pv = []
+    _product = product
+    for px, ex in factorint(p).items():
+        if a % px:
+            # `len(rx)` is at most 4
+            rx = _sqrt_mod_prime_power(a, px, ex)
         else:
-            res = _sqrt_mod_prime_power(a, p, 1)
-        if res:
-            if domain is ZZ:
-                for x in res:
-                    yield ZZ(x)
-            else:
-                for x in res:
-                    yield domain(x)
+            # `len(list(rx))` can be assumed to be large.
+            # The `itertools.product` is disadvantageous in terms of memory usage.
+            # It is also inferior to iproduct in speed if not all Cartesian products are needed.
+            rx = _sqrt_mod1(a, px, ex)
+            _product = iproduct
```

Comment:
Is there any case where `iproduct` can be disadvantageous?

Reply:
One thing I notice is that `iproduct` returns tuples. I am not sure how this impacts memory (creating lots of tuples) compared to the deleted `_product` routine which keeps using the same mutable list to return the result. Since one will consume these one at a time they can choose whether they want to convert that list to a tuple or just use it's values and let it be repopulated with values on the next call. I suspect it would be better to replace the `iproduct` with `_product`. @oscarbenjamin , do you have any insights here?

Reply:
I just checked `_product` vs `iproduct` and one of the undesirable aspects of `_product` is that it appears to try to consume one iterable before the next, so `_product(S.Integers, S.Integers)` generates `(0, 0), (0, 1), (0, -1), (0, 2), (0, -2), (0, 3), (0, -3), (0, 4), (0, -4),...` while `iproduct` generates `(0, 0), (1, 0), (0, 1), (-1, 0), (1, 1), (0, -1), (2, 0), (-1, 1), (1, -1),...`. Unless memory usage is an issue, this feature alone makes `iterables` version better.

Reply:
(And if memory use is an issue, that can be fixed in another PR.)

Reply:
> I am not sure how this impacts memory (creating lots of tuples) compared to the deleted `_product` routine which keeps using the same mutable list to return the result.

I doubt that the memory usage is significant. In CPython a tuple consumes less memory than the objects that it contains (if there are no repeated objects).

Reply:
@sylee957 

As shown below, `iproduct` is faster when only one element is to be retrieved.

```python
%%timeit
next(product(range(1000), range(1000)))
28.6 µs ± 4.16 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

%%timeit
next(iproduct(range(1000), range(1000)))
4.76 µs ± 95.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

On the other hand, `product` is faster when enumerating all elements.

```python
%%timeit
list(product(range(1000), range(1000)))
166 ms ± 28.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%%timeit
list(iproduct(range(1000), range(1000)))
361 ms ± 2.45 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```